src/app/AttributePathExpandIterator.h

```diff
@@ -20,78 +20,100 @@
- * AttributePathParams with wildcards. For AttributePathParams with a concrete path (i.e. does not contain wildcards),
- * AttributePathExpandIterator will emit them as-is.
- *
- * The typical use of AttributePathExpandIterator may look like:
- * ConcreteAttributePath path;
- * for (AttributePathExpandIterator iterator(AttributePathParams); iterator.Get(path); iterator.Next()) {...}
- *
- * The iterator does not copy the given AttributePathParams. The given AttributePathParams must remain valid when using the
- * iterator. If the set of endpoints, clusters, or attributes that are supported changes, AttributePathExpandIterator must be
- * reinitialized.
- *
- * A initialized iterator will return the first valid path, no need to call Next() before calling Get() for the first time.
- *
- * Note: Next() and Get() are two separate operations by design since a possible call of this iterator might be:
- * - Get()
- * - Chunk full, return
- * - In a new chunk, Get()
- *
- * TODO: The AttributePathParams may support a group id, the iterator should be able to call group data provider to expand the group
- * id.
- */
+/// Handles attribute path expansions
+/// Usage:
+///
+/// - Start iterating by creating an iteration state
+///
+///      mState =  AttributePathExpandIterator::State::StartIterating(path);
+///
+/// - Use the iteration state in a for loop:
+///
+///      ConcreteAttributePath path;
+///      for (AttributePathExpandIterator iterator(mState); iterator->Next(path);) {
+///         // use `path` here`
+///      }
+///
+///   OR:
+///
+///      ConcreteAttributePath path;
+///      AttributePathExpandIterator iterator(mState);
+///
+///      while (iterator.Next(path)) {
+///         // use `path` here`
+///      }
+///
+/// USAGE requirements and assumptions:
+///
+///    - There should be only one single AttributePathExpandIterator for a state  at a time.
+///
+///    - `State` is automatically updated by the AttributePathExpandIterator, so
+///      calling `Next` on the iterator will update the state variable.
+///
+///
 class AttributePathExpandIterator
 {
 public:
-    AttributePathExpandIterator(DataModel::Provider * provider, SingleLinkedListNode<AttributePathParams> * attributePath);
-
-    /**
-     * Proceed the iterator to the next attribute path in the given cluster info.
-     *
-     * Returns false if AttributePathExpandIteratorDataModeDataModel has exhausted all paths in the given AttributePathParams list.
-     */
-    bool Next();
-
-    /**
-     * Fills the aPath with the path the iterator currently points to.
-     * Returns false if the iterator is not pointing to a valid path (i.e. it has exhausted the cluster info).
-     */
-    bool Get(ConcreteAttributePath & aPath)
+    class State
     {
-        aPath = mOutputPath;
-        return (mpAttributePath != nullptr);
-    }
+    public:
+        friend class AttributePathExpandIterator;
```

Comment:
Why is the friendship needed?

Reply:
`AdvanceOutputPath` handles direct access to mAttributePath/mLastOutputPath.

Reply:
I think this could have been solved by using a struct with private members, rather than friend. Let's leave it for now.